packages/flutter_service_worker/lib/flutter_service_worker.dart

```diff
@@ -0,0 +1,60 @@
+// Copyright 2020 The Chromium Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'src/_io_impl.dart' if (dart.library.js) 'src/_web_impl.dart';
+
+/// An API for interacting with the service worker for application caching and
+/// installation.
+///
+/// On platforms other than the web, this delegates to a no-op implementation.
+///
+/// See also:
+///
+///  * https://web.dev/customize-install/
+abstract class ServiceWorkerApi {
+  /// Initialize the service worker API.
+  ///
+  /// This method should be called immediately in main, before calling
+  /// [runApp].
+  void init();
+
+  /// A future that resolves when it is safe to call [showInstallPrompt].
+  ///
+  /// If the application is not compatible with a service worker installation,
+  /// for example by running on http instead of https, then this future
+  /// will never resolve.
+  ///
+  /// This installation prompt event is currently only supported on Chrome.
+  /// On other browsers this future will never resolve.
+  ///
+  /// Not all browsers
+  Future<void> get installPromptReady;
+
+  /// Trigger a prompt that allows users to install their application to the
+  /// device/home screen location.
+  ///
+  /// Returns a boolean that indicates whether the installation prompt was
+  /// accepted.
+  ///
+  /// This installation prompt event is currently only supported on Chrome.
+  /// On other browsers [installPromptReady] will never resolve.
+  ///
+  /// Throws a [StateError] if this function is called before [installPromptReady]
```

Comment:
Consider an API that doesn't give the developer an opportunity to misuse it (instead of the StateError exception). For example, the `installPromptReady` future could return an object that has the `showInstallPrompt` method. Strawman:

```dart
serviceWorkerApi.installPromptReady.then((WepAppInstaller installer) async {
  await installer.showInstallPrompt();
});
```

This way there's no way to call the `showInstallPrompt` inappropriately.